### PR TITLE
Function as data

### PR DESCRIPTION
**Pull request by thomasp85**
This is a follow up on PR #1486

Summary of changes:

- Add `fortify.function` that returns function unchanged
- Add `layer_data` method to Layer base class that takes the plot data as input and returns the data associated with the layer
- Remove the need for waiver-check in `map_layout`
- Add unit tests for `layer_data` method

Possible additions:
- Documentation of function-as-data functionality somewhere (don't really know the best location for this - any idea?)

### CODE REVIEW

File: R/layer.r [hunk 1]
```diff
@@ -101,6 +101,19 @@ Layer <- ggproto("Layer", NULL,
...
+  layer_data = function(self, plot_data) {
+    if (is.waive(self$data)) {
+      return(plot_data)
+    } else if (is.function(self$data)) {
+      data <- self$data(plot_data)
+      if (!is.data.frame(data)) {
+        stop("Data function must return a data.frame")
+      }
+      return(data)
+    }
```
Comment (hadley) [inline R/layer.r]: I think this would be better with no returns and a final else block

File: R/layer.r [hunk 2]
```diff
@@ -101,6 +101,19 @@ Layer <- ggproto("Layer", NULL,
     cat(snakeize(class(self$position)[[1]]), "\n")
   },
 
+  layer_data = function(self, plot_data) {
+    if (is.waive(self$data)) {
+      return(plot_data)
+    } else if (is.function(self$data)) {
+      data <- self$data(plot_data)
+      if (!is.data.frame(data)) {
+        stop("Data function must return a data.frame")
```
Comment (hadley) [inline R/layer.r]: This needs `call. = FALSE`

File: tests/testthat/test-layer.r [hunk 3]
```diff
@@ -32,3 +32,16 @@ test_that("strip_dots remove dots around calculated aesthetics", {
...
+
+test_that("layer_data returns a data.frame", {
+  l <- geom_point()
+  expect_equal(l$layer_data(mtcars), mtcars)
+  l <- geom_point(data = head(mtcars))
+  expect_equal(l$layer_data(mtcars), head(mtcars))
+  l <- geom_point(data = head)
+  expect_equal(l$layer_data(mtcars), head(mtcars))
+  l <- geom_point(data = nrow)
+  expect_error(l$layer_data(mtcars))
```
Comment (hadley) [inline tests/testthat/test-layer.r]: Can you please check for the specific error message?

**Comment by hadley** [comment]
This is looking good.  A few points:

* Can you please add a bullet point to NEWS?
* Probably best to document this in `layer()`

**Comment by thomasp85** [comment]
All points have been addressed...

File: R/layer.r [hunk 4]
```diff
@@ -101,6 +128,20 @@ Layer <- ggproto("Layer", NULL,
     cat(snakeize(class(self$position)[[1]]), "\n")
   },
 
+  layer_data = function(self, plot_data) {
+    if (is.waive(self$data)) {
+      data <- plot_data
```
Comment (hadley) [inline R/layer.r]: You can drop the `data <-` in all the list cases

**Comment by hadley** [comment]
One other idea now that I've merged it - what if the fortify happened in `layer_data()` as well?

**Comment by thomasp85** [comment]
Possible error because of unknown data type will first happen when plotting - don't know if that is desirable?

**Comment by thomasp85** [comment]
Should I redo the layer_data or is it merged and accepted and it was just a point for the future?

**Comment by hadley** [comment]
Just a point for the future. I fixed locally

**Pull request merged**